### ISSUE TITLE
Transactional counter problem

### ISSUE THREAD

**Issue by PeterKnego** [orientechnologies/orientdb#3457]
I'm evaluating several NoSQL databases, including OrientDB. I'm a OrientDB newbie, trying to write a multi-threaded test that stresses OrientDB transactions.

It's a simple standalone test that uses a number of parallel threads that transactionally increment a single counter: read an entity, increases the value of number field and them save the entity, all inside a transaction.

It creates a lot of contention on one entity - that's the whole point. It kind of works, detects contention most of the time and properly rolls back the transaction. But it also fails: it sometimes does not detect "dirty" writes. 

I'm testing this on a locally installed OrientDB 2.0-rc2, client connects to it via a remote protocol.

Any ideas why sometimes transaction collisions get detected, while sometimes they pass?

```
public class OrientDbSimpleTransact {

	private static final AtomicInteger verifyCounter = new AtomicInteger(0);
	private static final AtomicInteger collisions = new AtomicInteger(0);

	public static void main(String[] args) throws InterruptedException {

		ODatabaseDocument db = new ODatabaseDocumentTx("remote:localhost/test")
				.open("admin", "admin");
		OSchema schema = db.getMetadata().getSchema();

		System.out.println("MVCC: " + db.isMVCC());

		String tableName = "Counter";
		int threads = 5;
		int repeat = 20;

		// register class
		OClass cls = schema.getOrCreateClass(tableName);
		String fieldName = "number";
		if (!cls.existsProperty(fieldName)) {
			cls.createProperty(fieldName, OType.INTEGER);
		}

		// create counter entity
		ODocument doc = new ODocument(tableName);
		doc.field(fieldName, 0);
		ORID orid = db.save(doc).getIdentity();

		// starting threads
		ExecutorService executor = Executors.newFixedThreadPool(100);
		Collection<Callable<Void>> tasks = new ArrayList<>();
		for (int n = 1; n <= threads; n++) {

			int delta = 1;
			tasks.add(new CounterIncrement(orid, repeat));
			System.out.println("Added task:" + n + " delta:" + delta);
		}
		executor.invokeAll(tasks);
		executor.shutdown();

		// load the counter again
		ODocument res = db.load(orid, null, true);

		System.out.println("Counter updates: " + res.toMap().get(fieldName));
		System.out.println("Verify counter: " + verifyCounter);
		System.out.println("Collisions: " + collisions);
		if (res.toMap().get(fieldName) != verifyCounter) {
			System.out.println("Error: number of updates (" +
					res.toMap().get(fieldName) + ") is not equal to verify counter (" + verifyCounter + ").");
		}
		db.close();

	}

	public static class CounterIncrement implements Callable<Void> {

		public CounterIncrement(ORID key, int repeat) {
			this.key = key;
			this.repeat = repeat;
		}

		private ORID key;
		private int repeat;

		/**
		 * Transactionally increments the counter
		 */
		@Override
		public Void call() throws Exception {

			ODatabaseDocument db = new ODatabaseDocumentTx("remote:localhost/test")
					.open("admin", "admin");
			while (repeat != 0) {

				try {
					db.begin(OTransaction.TXTYPE.OPTIMISTIC);
					ODocument entity = db.load(key, null, true);

					if (entity != null) {

						// increment field 'number' by one
						int val = entity.field("number");
						val++;
						entity.field("number", val);

						db.save(entity);
						db.commit();
						verifyCounter.addAndGet(1);

						// repeat variable is only decreased when transaction is committed successfully
						repeat--;
						System.out.println("Updated entity:" + key + " number:" + entity.field("number"));
					} else {
						System.out.println("transact " + Thread.currentThread().getName() + " Not found! key=" + key);
						db.rollback();
					}
				} catch (RuntimeException re) {
					collisions.addAndGet(1);
					System.out.println("Collision: " + re.getMessage());
					db.rollback();
				}
			}
			db.close();

			return null;
		}
	}
}
```

**Comment by lvca** [comment]
Have you tried final 2.0?

**Comment by PeterKnego** [comment]
Yes, I just tried it with `OrientDB Server v2.0 (build UNKNOWN@r; 2015-01-20 16:41:17+0000)`. I get the same error.

I'm a new with OrientDB - are there any settings or config that I'm getting wrong? Also, I'm creating a new `ODatabaseDocumentTx` for every thread as I did not succeed to create a [multi-threaded custom database factory](http://www.orientechnologies.com/docs/last/orientdb.wiki/Java-Multi-Threading.html#custom-database-factory).
